core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,445 @@
+    }
+    Expression oldFilter = null, newFilter = null;
+    if (join.getPostJoinFilter().isPresent()) {
+      newFilter = visitExpression(oldFilter = join.getPostJoinFilter().get());
+    }
+    if (oldLeft == newLeft
+        && oldRight == newRight
+        && oldCondition == newCondition
+        && oldFilter == newFilter) {
+      return join;
+    }
+    ImmutableJoin.Builder builder =
+        ImmutableJoin.builder().from(join).left(newLeft).right(newRight);
+    if (join.getCondition().isPresent()) {
+      builder.condition(newCondition);
+    }
+    if (join.getPostJoinFilter().isPresent()) {
+      builder.postJoinFilter(newFilter);
+    }
+    return builder.build();
+  }
+
+  @Override
+  public NamedScan visit(NamedScan namedScan) {
+    return namedScan;
+  }
+
+  private List<Expression> copyOnWrite(List<Expression> oldExpressions) {
+    int expressionCount = oldExpressions.size();
+    List<Expression> newExpressions = oldExpressions;
+    for (int i = 0; i < expressionCount; i++) {
+      Expression oldExpr = oldExpressions.get(i);
+      Expression newExpr = visitExpression(oldExpr);
+      if (oldExpr != newExpr) {
+        if (oldExpressions == newExpressions) {
+          newExpressions = new CopyOnWriteArrayList<>(oldExpressions);
+        }
+        newExpressions.set(i, newExpr);
+      }
+    }
+    return newExpressions;
+  }
+
+  @Override
+  public Project visit(Project project) {
+    Rel oldInput = project.getInput();
+    Rel newInput = oldInput.accept(this);
+    List<Expression> oldExpressions = project.getExpressions();
+    List<Expression> newExpressions = copyOnWrite(oldExpressions);
+    return oldInput == newInput && oldExpressions == newExpressions
+        ? project
+        : ImmutableProject.builder()
+            .from(project)
+            .input(newInput)
+            .expressions(newExpressions)
+            .build();
+  }
+
+  @Override
+  public Sort visit(Sort sort) {
+    Rel oldInput = sort.getInput();
+    Rel newInput = oldInput.accept(this);
+    return oldInput == newInput ? sort : ImmutableSort.builder().from(sort).input(newInput).build();
+  }
+
+  @Override
+  public VirtualTableScan visit(VirtualTableScan virtualTableScan) {
+    return virtualTableScan;
+  }
+
+  private Expression visitExpression(Expression expression) {
+    ExpressionVisitor<Expression, RuntimeException> visitor =
+        new ExpressionVisitor<Expression, RuntimeException>() {
+
+          public interface Getter<T> {
+            public Expression get(T element);
+          }
+
+          public interface Builder<T> {
+            public T build(T element, Expression e);
+          }
+
+          private static <T> List<T> copyOnWrite(
+              ExpressionVisitor visitor, List<T> oldExpressions) {
+            return copyOnWrite(
+                visitor, oldExpressions, element -> (Expression) element, (element, e) -> (T) e);
+          }
+
+          private static <T> List<T> copyOnWrite(
```

Comment:
I don't understand the purpose of this method. If we use the patterns proposed above, doesn't this method go away?

Reply:
This method uses two lambdas - one to get the expression from an item and the other to build an item from an expression. If item is an expression, then the lambdas are just passthroughs.